Salt'ify mark-control-plane phase

gdemonet opened the pull request:
Note that we also added an annotation for specifying the CRI socket used, as can be seen in the [kubeadm dry-run output](https://github.com/NicolasT/kubeadm-dry-run/blob/master/kubeadm-dry-run.txt#L518-L524).

Fixes GH-582.

gdemonet committed:
Add `Node` annotations management in `kubernetes` state

Issue: GH-582

gdemonet committed:
Add `Node` taints management in `kubernetes` state

Issue: GH-582

gdemonet committed:
Add `kubeadm.init.mark-control-plane` step

Issue: GH-582

NicolasT commented:
@bert-e wait

NicolasT commented:
Let's wait for CI to merge #689, then rebase this one on `development/2.0` and retarget this PR to `development/2.0` (no need to close this PR, can edit it in-place).

NicolasT commented on salt/_modules/kubernetes.py:
```diff
@@ -1797,3 +1797,132 @@ def replace_role(
+    finally:
+        _cleanup(**cfg)
+
+    return None
```
Is this done in the other procedures? There should be no need, either `return api_response` is reached, or the `return None` or `raise` in the `except` clause, or some other exception bubbles up. This is unreachable.

Also applies to other occurrences.

gdemonet replied:
Indeed, not sure exactly why I added them, will remove when rebasing.

NicolasT commented on salt/_modules/kubernetes.py:
```diff
@@ -1797,3 +1797,132 @@ def replace_role(
+    if match is not None:
+        return match['spec']['taints']
+
+    return None
```
This means the node doesn't exist, rather than 'no taints'. If this is how other procedures in the module work, then fine. We should rework it anyway :stuck_out_tongue:

gdemonet replied:
Actually I should have returned an empty list instead (copy-pasted the None as for the other methods I guess), but yes this module needs a rework :)

NicolasT commented on salt/_states/kubernetes.py:
```diff
@@ -1009,6 +1009,179 @@ def node_label_present(
+            annotation_value=value,
+            node_name=node,
+            **kwargs
+        )
```
In this code-path, `ret['comment']` is not set.

gdemonet replied:
Done.

NicolasT commented on salt/_states/kubernetes.py:
```diff
@@ -1009,6 +1009,179 @@ def node_label_present(
+
+def node_taints_present(name, taints, **kwargs):
+    '''
+    # TODO: describe and provide expls
```
I'd prefer no docstring (which can be caught by static analysis) than a docstring which is de-facto empty, no?

gdemonet replied:
Agreed, I was hoping to find the courage to fill it before pushing...

NicolasT commented on salt/_states/kubernetes.py:
```diff
@@ -1009,6 +1009,179 @@ def node_label_present(
+
+    existing_taints = {
+        t['key']: t
+        for t in __salt__['kubernetes.node_taints'](name, **kwargs)
```
Prefer not to call things with side-effects in a comprehension.

gdemonet replied:
Split the comprehension in two.

NicolasT commented on salt/_states/kubernetes.py:
```diff
@@ -1009,6 +1009,179 @@ def node_label_present(
+        **kwargs
+    )
+
+    ret['result'] = True
```
`ret['comment']` not set

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/kubeadm/init/mark-control-plane/configured.sls:
```diff
@@ -0,0 +1,31 @@
+Apply control-plane role label:
+  kubernetes.node_label_present:
+    - node: {{ hostname }}
+    - name: "node-role.kubernetes.io/master"
```
@TeddyAndrieux prefers `name` to go first, then `kubeconfig` and `context`, then `node` and `value`. I kinda agree (had to be convinced though ;))

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/kubeadm/init/mark-control-plane/configured.sls:
```diff
@@ -0,0 +1,31 @@
+    - taints:
+        - key: "node-role.kubernetes.io/master"
+          effect: "NoSchedule"
+    - kubeconfig: {{ kubeconfig }}
```
Same

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/kubeadm/init/mark-control-plane/configured.sls:
```diff
@@ -0,0 +1,31 @@
+  kubernetes.node_annotation_present:
+    - node: {{ hostname }}
+    - name: "kubeadm.alpha.kubernetes.io/cri-socket"
+    # NOTE: the socket location may become configurable, we should make sure
```
`TODO` i.s.o. `NOTE`.

gdemonet replied:
Done.

NicolasT reviewed (changes requested):
Definitely requires rebase after #685 lands.

gdemonet committed:
Clean state execution comments

These were not always properly set

Issue: GH-582

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

gdemonet commented:
Rebased and integrated review comments from @NicolasT

gdemonet commented:
Build failed since `mark-control-plane` runs before `addons`, and as such `python2-kubernetes` is not yet installed when we execute this step.

A quick fix could be to include `..addons.dependencies` in `metalk8s.kubeadm.init.mark-control-plane.init`, but this doesn't look "clean".
I'm wondering if doing something like a `metalk8s.python-kubernetes` formula could be useful / acceptable.

gdemonet committed:
Add formula for managing python-kubernetes client

Since we need this client in multiple kubeadm phases, we extract its
installation in a dedicated formula instead of leaving it as a
`dependencies` formula for the `addons` phase.

Issue: GH-582

cmonfort commented on salt/_modules/kubernetes.py:
```diff
@@ -1975,3 +1975,126 @@ def replace_daemonset(
+        body = {
+            'metadata': {
+                'annotations': {
+                    annotation_name: None}
```
Is the annotation removed when its value is set to `None`?
If not, it's not really a removal, and if I get it afterwards, I will not have a failure because it does not exist, but retrieve a `None` value...)

gdemonet replied:
Yup, setting an annotation or label to `None` in a PATCH call will effectively remove it - setting a label/annotation with _no value_ would be done by using an empty string instead.

cmonfort commented on salt/_states/kubernetes.py:
```diff
@@ -1009,6 +1009,171 @@ def node_label_present(
+            **kwargs
+        )
+
+    old_annotations = copy.copy(annotations)
```
Get the full annotations block? not only the value of one we are changing?

gdemonet replied:
Same comment as for the signature of `node_annotations`, I copy-pasted what was done for labels, and they indeed show the whole labels block instead of simply the one changed... To discuss in future improvements.

cmonfort commented on salt/_modules/kubernetes.py:
```diff
@@ -1975,3 +1975,126 @@ def replace_daemonset(
         _cleanup(**cfg)
+
+
+def node_annotations(name, **kwargs):
```
For consistency with other annotations commands, `name` parameter should be `node_name`.

gdemonet replied:
I copy-pasted the signature of [`node_labels`](https://github.com/scality/metalk8s/blob/020703de54117108286831e813d3076ecdf39d3a/salt/_modules/kubernetes.py#L321), I'll keep it as is until we refactor this custom module (which is essentially a long list of copy-pasted functions)

NicolasT reviewed (dismissed):
LGTM (didn't run)

gdemonet commented:
@bert-e approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/090c5f9d1fb98d37bafa44fd91d55bb6b6b205f7) did not succeed in branch feature/582-salt-mark-control-plane.




*The following options are set:* **approve**

ramanduh commented on salt/metalk8s/kubeadm/init/mark-control-plane/configured.sls:
```diff
@@ -0,0 +1,31 @@
+    - kubeconfig: {{ kubeconfig }}
+    - context: {{ context }}
+    - node: {{ hostname }}
+    # TODO: the socket location may become configurable, we should make sure
```
I think, it's already configurable in default.yaml: `kubelet.service.options.container-runtime-endpoint`

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
